### PR TITLE
Update i3 to 4.20.1, replace to xz files

### PR DESCRIPTION
**Pull request by anadahz**
Since i3 version 4.19 [1] the releases are release as an xz archive. [1] https://i3wm.org/downloads/

### CODE REVIEW

**Comment by holiman** [comment]
This PR was opened on Sep 11, 2022. Ten days later, on Sep 21, version `4.21` was released: https://i3wm.org/downloads/ . Perhaps worth updating to that one?

**Comment by holiman** [comment]
The pipeline fails, and I think it's due to the fetch-step. The previous last successfull merged PR job is here: https://gitlab.com/QubesOS/qubes-desktop-linux-i3/-/jobs/2699089150/artifacts/browse/artifacts/distfiles/ . It fetched the `i3-4.18/2.tar.bz2` into 
```
artifacts/distfiles/i3-4.18.2.tar.bz2
```
And then it was available for the build phase. 

The failing CI job is here: https://gitlab.com/QubesOS/qubes-desktop-linux-i3/-/pipelines/667274307

The dom0 build job is failing, seemingly because the i3wm source artefact is missing. The artefacts from the `r.4.2:prep_sources` job does not include a corresponding source-ball as the previous job:
https://gitlab.com/QubesOS/qubes-desktop-linux-i3/-/jobs/3176017418/artifacts/browse/artifacts/distfiles/

Log from when it worked:
```
21:23:32,750 [executor:docker:56d8e4bac4] output: + mv untrusted_i3-4.18.2.tar.bz2 /builder/desktop-linux-i3/i3-4.18.2.tar.bz2
```
Log from when it failed: 
```
13:17:28,159 [executor:local:/tmp/139842906404224197c2b45/builder] output: + mv /home/gitlab-runner/qubes-builder/artifacts/tmp/tmpsr2bxw1e/untrusted_i3-4.21.tar.xz /home/gitlab-runner/qubes-builder/artifacts/tmp/tmpsr2bxw1e/i3-4.21.tar.xz
```
I think the `mv` step when it moves the source-bundle from untrusted might not be copying it to the correct location... (?)

**Comment by marmarek** [comment]
The hash file name is wrong - should be .sha512, not .asc.sha512.

File: .qubesbuilder [hunk 1]
```diff
@@ -14,5 +14,5 @@ vm:
       - '@PLUGINS_DIR@/source_deb/scripts/debian-quilt @SOURCE_DIR@/series-debian.conf @BUILD_DIR@/debian/patches'
 source:
   files:
-  - url: https://i3wm.org/downloads/i3-4.18.2.tar.bz2
-    sha512: i3-4.18.2.tar.bz2.sha512
+  - url: https://i3wm.org/downloads/i3-4.21.tar.xz
+    sha512: i3-4.21.tar.xz.asc.sha512
```
Comment (marmarek) [inline .qubesbuilder]: Here too.

**Comment by holiman** [comment]
Pipeline got further this time: 
```
tar jxvf qubes-src/desktop-linux-i3/i3-4.21.tar.xz -C /home/gitlab-runner/qubes-builder/chroot-dom0-fc32//home/user/qubes-src/desktop-linux-i3/debian-pkg --strip-components=1
bzip2: (stdin) is not a bzip2 file.
tar: Child returned status 2
tar: Error is not recoverable: exiting now
make[1]: Leaving directory '/home/gitlab-runner/qubes-builder'
make[1]: *** [qubes-src/desktop-linux-i3/Makefile.builder:16: source-debian-copy-in] Error 2
make: *** [Makefile:273: desktop-linux-i3-dom0] Error 1
```

tar args: 
```
  -j, --bzip2                filter the archive through bzip2
  -J, --xz                   filter the archive through xz
```
So I think maybe `tar jxvf $(SRC_FILE) -C $(CHROOT_DIR)/$(DIST_SRC)/debian-pkg --strip-components=1` in Makefile.builder should become `tar Jxvf $(SRC_FILE) -C $(CHROOT_DIR)/$(DIST_SRC)/debian-pkg --strip-components=1` ...(?)

**Comment by Minimalist73** [comment]
About the last pipeline failure:

The `0001-Show-qubes-domain-in-configurable-colored-borders.patch` file need to be fully updated with the latest i3 code changes. I spent some time around it and made my own patch which went through but I also needed to update the build method since they switched to meson + ninja (so you need to add them to the build spec file as requirement). I deleted the 2 other patch since they are not required anymore (widgets disappearing is fixed since a few versions and the other one is based on the previous build method). I have a lot of build errors for now and I don't have really much time to inspect that more and edit the patch accordingly.

**Comment by holiman** [comment]
I tested building i3wm `4.21` on a fedora. I didn't use the qubes-builder flow, because I haven't quite figured out how it works yet. 
As for dependencies, these seem to be required: 
```
sudo dnf install  startup-notification-devel \
 xcb-util-devel xcb-util-cursor-devel xcb-util-keysyms-devel xcb-util-wm-devel xcb-util-xrm-devel\ 
 libxkbcommon-devel libxkbcommon-x11-devel \
 yajl-devel pcre2-devel \
 cairo-devel pango-devel \
 libev-devel
```
Afaict those are covered by these requirements: https://github.com/QubesOS/qubes-desktop-linux-i3/blob/master/i3.spec.in#L20 . Might even be the case that some deps can be removed (?). 

And the build itself requires `meson` and `ninja`. 
```
pip3 install --user meson
```
For `ninja`, I grabbed a binary from https://github.com/ninja-build/ninja/releases/tag/v1.11.1. Building it is two steps, configure/setup via `meson` and do the actual compilation using `ninja`:  

```
$ mkdir build && cd build
$ meson .. && ninja
```

The patch does not apply cleanly. I uploaded the 4.21 sources as a repo to https://github.com/holiman/i3wm-temp. The first commit are the raw sources. The second commit is the first attempt to manually apply the `0001`-patch. The third commit is a fixup, making it compile again. 

I applied the patch-changes, but I did it without a lot of consideration, so there's a possibility I did something wrong. The patches were pretty easy to apply, I think the only "new" change that was needed was the change to `tiling_drag.c` . 

Anyway, a clean`0001`-patch can be extracted from commit 2 and 3 in that repo. I'd appreciate if someone more familiar with the i3wm internals could sanity-check my patchset. [Github diff on the two commits here](https://github.com/holiman/i3wm-temp/compare/7d48c13e36a84e99a0e760545833032bb0d25a14...037a8e3f8402734919797e4162cef824dc3cefbf)

New diff: [0001-Show-qubes-domain-in-configurable-colored-borders.new.txt](https://github.com/QubesOS/qubes-desktop-linux-i3/files/9808510/0001-Show-qubes-domain-in-configurable-colored-borders.new.txt)

**Comment by holiman** [comment]
Got a bit further on this work over at https://github.com/QubesOS/qubes-desktop-linux-i3/pull/50 . The dom0 build is now working

**Pull request merged**